### DISCUSSION TITLE
How to use JS (or even TS) test-script file instead of Yaml

### DISCUSSION THREAD

**Discussion by HugoImaios** [artilleryio/artillery#2566] [Artillery]
Through [the documentation](https://www.artillery.io/docs/get-started/first-test#create-an-artillery-test-script), we can find this : 
> Artillery test scripts are usually written as YAML, but they can also be written in JavaScript.

Which sounds great :heart_eyes:  !
After exploring a bit the [github packages](https://github.com/artilleryio/artillery/tree/main/packages), we can even find [this one](https://github.com/artilleryio/artillery/tree/main/packages/types) which is published on npm as `@artilleryio/types` and which seems to provides type definitions for artillery config, scenarios, etc. :tada:  _(subsidiary question : why is there a DT package (@types/artillery) then ?)_

**But :** how can we run this using artillery ? `artillery run` seems to be locked onto yaml files, causing ```YAMLException: end of the stream or a document separator is expected```

**Comment by bernardobridge** [comment]
Hello @HugoImaios 👋,

To use Typescript support, you still write your basic skeleton of a script in YAML. However, in the YAML file you can point to a Typescript processor which contains exported functions, so your actual logic can be written in Typescript or Javascript. How to load the several different types of processors (.js, .mjs or .ts) is explained [here](https://www.artillery.io/docs/reference/test-script#processor---load-custom-code).

From there, you would write functions you can then use in your scenarios. 
- Here is an example for [Playwright](https://www.artillery.io/docs/reference/engines/playwright#example-3-using-typescript). 
- For the HTTP engine, if you wanted to write in pure Javascript/Typescript code, you can also call [functions](https://www.artillery.io/docs/reference/engines/http#function-steps-in-a-flow) within the YAML (maybe a single one in your case), but you'll need to emit your own [custom metrics](https://www.artillery.io/docs/reference/extension-apis#tracking-custom-metrics), as you won't have the default HTTP engine metrics emitted then - only the [default ones](https://www.artillery.io/docs/reference/reported-metrics#metrics-reported-by-artillery) would be emitted.

---

As for your other questions:

> we can even find [this one](https://github.com/artilleryio/artillery/tree/main/packages/types) which is published on npm as @artilleryio/types and which seems to provides type definitions for artillery config, scenarios, etc.

The name of that package is probably misleading, and we will likely change it soon. At the moment, `@artilleryio/types` just powers a JSON schema used by our [VS Code Extension](https://www.artillery.io/docs/resources/vs-code-extension). The type definitions there are an artifact of a legacy implementation.

>subsidiary question : why is there a DT package (@types/artillery) then ?

Those types are not published by the Artillery team, but by an open source contributor. It is mostly quite good, though!

**Comment by HugoImaios** [comment]
hello @bernardobridge !
Thanks again for the work done on the [Typescript and Playwright support](https://github.com/artilleryio/artillery/discussions/1623), I'm testing it (lol) theses days and it's very nice.
I think I was not clear in my first message. I was not talking about the processors scripts, but really about what is called "test scripts" and which contains the `config` and `scenarios` keys. By the doc I quoted, I naively though that this yaml file could be written in js/ts.
I now get that it is not possible (but it's ok, it might not be a huge use-case) :)

**Reply by bernardobridge** [reply]
Ah I see, sorry if I misunderstood you! I'll update the docs soon to make this a bit more clear, though!

We will probably want to support running from pure JS files at some point, but it's not something we have an ETA for right now. At the moment, as you can run `ESM` (via .mjs), `Typescript` (via .ts) and `Javascript` (.js), it covers the most common use cases with the workaround I described above, as the logic would still be written outside of the YAML, and the YAML would mostly serve as an entrypoint for config. 

Do let us know if you run into any issues though!

---

P.S. Glad to know Typescript support is working well for you! 🎉

**Reply by HugoImaios** [reply]
you're totally right.
If this ever comes back to the table, here is our use-case :
we have playwright configuration (mostly `contextOptions` and mostly `extraHTTPHeaders`) logic based on env vars, which we implemented in our `playwright.config.ts`. For artillery, we currently have to re-implement this in the good old artillery.yaml file, mostly through the `environments` key

**Reply by HugoImaios** [reply]
+usecase for TS would be to get auto-completion for artillery config/scenarios

**Reply by bernardobridge** [reply]
Ahh! That's interesting to know actually, hadn't considered that 🤔.

We'll definitely keep this in mind for further updates to the Playwright engine!

**Reply by bernardobridge** [reply]
> +usecase for TS would be to get auto-completion for artillery config/scenarios

Yes - we are definitely interested in either contributing ourselves to the `@types/artillery` or publishing our own official ones. It's not something we have gotten to yet, but it will come! 

Thanks for all the feedback.

**Comment by electrovir** [comment]
Just wanted to throw my 2¢ in here @bernardobridge. I recently discovered Artillery and was super interested in it all up until I found out that test cases _must_ be written in yaml. No type safety, no code reuse, white-space as syntax... that's not something they my team can rely on 🙃

(We have to use yaml already for GitHub Actions and that's bad enough as it is.)

**Reply by bernardobridge** [reply]
Hi @electrovir 👋 . Thanks for the feedback, it's helpful to know that! It does go a bit both ways, as we do have many users who prefer the YAML format (it depends how complex and logic-heavy your tests are, I suppose).

As mentioned above, we did introduce initial Typescript support recently, and you can pretty much write your whole logic in Typescript, and the YAML will just serve as configuration (pointing to the Typescript function needed). Does that not work for your use case?

In the near future we'll be looking to make this more robust, so any feedback helps with that.

**Reply by electrovir** [reply]
It seems you still have to define all your scenarios in the yaml file though, is that not the case? I've also seen several warnings in the docs about TypeScript support being experimental, which naturally concerns me 😅

**Reply by bernardobridge** [reply]
@electrovir , the scenario will need to be specified in the YAML config, but not the actual code (you'd point at a function to run it instead). Playwright already works like that. I'll try to add an example soon for HTTP tests soon, but it would look something like this:

scenario.yml
```yml
config:
  target: https://example.com
  processor: "./flows.ts"
scenarios:
  - flow:
      - function: "getLoginPage"
```

flows.ts
```ts
//install an http client yourself
const axios = require('axios');

export async function getLoginPage(context, ee) {
    const startTime = Date.now();
    await axios.get(`${context.vars.target}/login`);
    const endTime = Date.now();

    const duration = endTime - startTime;
    ee.emit('histogram', 'http.response_time', duration);
};
```

You get the overhead of having to emit the metrics you care about yourself, but the [custom metrics API ](https://www.artillery.io/docs/reference/extension-apis#tracking-custom-metrics) allows you to do whatever you need.

Like I said, in the future we will likely add more convenience features around this support, but it's already possible to do most things with it.

---

>I've also seen several warnings in the docs about TypeScript support being experimental, which naturally concerns me 😅

That's understandable! It's a relatively new feature, and we need more feedback and understanding of usage patterns before we consider it stable. But we haven't had any major reported issues.

Thanks again for your input!

**Reply by electrovir** [reply]
Okay, that sounds more promising than I initially interpreted. Do you happen to have any larger example repos off-hand that you could point me to that extensively use TypeScript to run Artillery tests?

**Reply by benjaminkay93** [reply]
^ i too would find it helpful having a larger example set for TS Artillery tests 🙏

**Comment by hassy** [comment]
For anyone who comes across this discussion via search - we added support for writing Artillery + Playwright tests fully in TypeScript in #3436, which will be released in Artillery v2.0.22.

You can see what it looks like in this example in the PR: https://github.com/artilleryio/artillery/pull/3436/files#diff-4a2a2bed71ededd5abe2fbeb9e2e26008094657ea06c9d771c80c03a3bad6934

**Reply by HugoImaios** [reply]
Hello Hassy !
Great news ! Haven't land my hands on our artillery scripts since months, but I'll certainly give this update a try when I'll come back to it !
Have a nice nye :)

**Reply by barth12** [reply]
Hello, we are writing our tests in TS now and this is compiled to JS, that part works well. However, should we still use https://www.npmjs.com/package/@types/artillery for the type support while developing?
